src/video/rpi_cam.cc

```diff
@@ -40,36 +41,47 @@ RPiCamera::getOutputSize() const
 
 bool
 RPiCamera::readFrame(cv::Mat &outFrame)
+{
+    if (!readGreyscaleFrame(m_Frame)) {
+        return false;
+    }
+
+    // Make sure output frame is the right size and type
+    outFrame.create(72, 152, CV_8UC3);
+
+    // Convert to the correct cv::Mat type
+    m_Frame.convertTo(outFrame, CV_8UC3);
+    return true;
+}
+
+bool
+RPiCamera::readGreyscaleFrame(cv::Mat &outFrame)
 {
     unsigned char buffer[72 * 19];
 
     // Check we're connected
     BOB_ASSERT(m_Socket != INVALID_SOCKET);
 
     // Make sure output frame is the right size and type
-    outFrame.create(72, 152, CV_32FC1);
+    outFrame.create(72, 152, CV_8U);
 
     // get the most recent UDP frame (grayscale for now)
     while (recv(m_Socket, buffer, 72 * 19, 0) > 0) {
-        // fill in the outFrame
-        //std::cout << (int) buffer[0] << std::endl;
-        int j = 0;
+        /*
+         * Fill in the outFrame.
+         *
+         * NB: We might be able to std::copy the memory from the buffer, but I
+         * won't do this just yet in case it inverts the rows and columns or
+         * something... -- AD
+         */
         for (int i = 0; i < 72 * 19 - 1; ++i) {
-            outFrame.at<float>(i % 72, buffer[0] + floor(i / 72)) = float(buffer[j]) / 255.0f;
-            ++j;
+            outFrame.at<uchar>(i % 72, buffer[0] + floor(i / 72)) = buffer[i];
```

Comment:
I now understand what's going on but it's pretty horribly innefficient, how about:

```c++
// Loop through rows
for(int i = 0; i < 72; i++) {
    // Get row pointer, with buffer offset
    uint8_t *subRow = outFrame.ptr<uint8_t>(r) + buffer[0];

    // Copy sub-row of data from buffer into image
    std::copy_n(&buffer[i * 19], 19, subRow);
}
```

Reply:
Also I'm concerned that ``buffer[0]`` is used both as a column offset for the block of data contained in the frame **and** the first pixel - that doesn't seem right

Reply:
That does look better... we might even be able to ``std::copy_n`` the entire buffer in one go.

Reply:
And the ``buffer[0]`` thing is correct, though we should probably make the buffer one byte bigger instead